### PR TITLE
feat(timepicker): have up/down arrow keys control time selection

### PR DESCRIPTION
**Pull request by joshfriend**
Similar to #2316, but:
* has tests/docs.
* properly consumes up/down events with `preventDefault`.

### CODE REVIEW

File: src/timepicker/timepicker.js [hunk 1]
```diff
@@ -114,6 +120,39 @@ angular.module('ui.bootstrap.timepicker', [])
 
   };
 
+  // Respond on up/down arrowkeys
+  this.setupArrowkeyEvents = function( hoursInputEl, minutesInputEl ) {
+    hoursInputEl.bind('keydown', function(e) {
+      if ( e.which === 38 ) { // up
+        $scope.$apply(function() {
+          $scope.incrementHours();
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: This should be rewritten as
<pre>
$scope.incrementHours();
$scope.$apply();
</pre>

File: src/timepicker/timepicker.js [hunk 2]
```diff
@@ -114,6 +120,39 @@ angular.module('ui.bootstrap.timepicker', [])
...
+        });
+        e.preventDefault();
+      }
+      else if ( e.which === 40 ) { // down
+        $scope.$apply(function() {
+          $scope.decrementHours();
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: Similarly here.

File: src/timepicker/timepicker.js [hunk 3]
```diff
@@ -114,6 +120,39 @@ angular.module('ui.bootstrap.timepicker', [])
...
   };
 
+  // Respond on up/down arrowkeys
+  this.setupArrowkeyEvents = function( hoursInputEl, minutesInputEl ) {
+    hoursInputEl.bind('keydown', function(e) {
+      if ( e.which === 38 ) { // up
+        $scope.$apply(function() {
+          $scope.incrementHours();
+        });
+        e.preventDefault();
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: This should be at the top of the if conditional

File: src/timepicker/timepicker.js [hunk 4]
```diff
@@ -114,6 +120,39 @@ angular.module('ui.bootstrap.timepicker', [])
...
+      }
+      else if ( e.which === 40 ) { // down
+        $scope.$apply(function() {
+          $scope.decrementHours();
+        });
+        e.preventDefault();
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: Same here

File: src/timepicker/timepicker.js [hunk 5]
```diff
@@ -114,6 +120,39 @@ angular.module('ui.bootstrap.timepicker', [])
...
+      }
+    });
+
+    minutesInputEl.bind('keydown', function(e) {
+      if ( e.which === 38 ) { // up
+        $scope.$apply(function() {
+          $scope.incrementMinutes();
+        });
+        e.preventDefault();
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: Same with this if block

File: src/timepicker/timepicker.js [hunk 6]
```diff
@@ -114,6 +120,39 @@ angular.module('ui.bootstrap.timepicker', [])
...
+      }
+      else if ( e.which === 40 ) { // down
+        $scope.$apply(function() {
+          $scope.decrementMinutes();
+        });
+        e.preventDefault();
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: Same here.

File: src/timepicker/timepicker.js [hunk 7]
```diff
@@ -26,6 +27,11 @@ angular.module('ui.bootstrap.timepicker', [])
       this.setupMousewheelEvents( hoursInputEl, minutesInputEl );
     }
 
+    var arrowkeys = angular.isDefined($attrs.arrowkeys) ? $scope.$parent.$eval($attrs.arrowkeys) : timepickerConfig.arrowkeys;
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: This would be better changed to use the isolate scope in the directive.

Comment (joshfriend) [inline src/timepicker/timepicker.js]: meaning I should use `$scope.$eval` instead of `$scope.$parent.$eval`?

File: src/timepicker/timepicker.js [hunk 8]
```diff
@@ -26,6 +27,11 @@ angular.module('ui.bootstrap.timepicker', [])
       this.setupMousewheelEvents( hoursInputEl, minutesInputEl );
     }
 
+    var arrowkeys = angular.isDefined($attrs.arrowkeys) ? $scope.$eval($attrs.arrowkeys) : timepickerConfig.arrowkeys;
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: I meant modify the directive scope to set `arrowkeys: '=?'`.  It should be present by the time the init function is run, so you can directly pull the value from `$scope.arrowkeys` in that case.

Comment (wesleycho) [inline src/timepicker/timepicker.js]: Actually upon looking at the rest of the source, change this back to $scope.$parent.$eval - I would rather have consistency with the rest of the source for now.

**Comment by wesleycho** [comment]
This LGTM once that change is made back to use $scope.$parent.$eval.

**Comment by joshfriend** [comment]
@wesleycho would you like me to rebase this branch so that 0027763 and 8eb93ce are squashed?

**Comment by wesleycho** [comment]
That won't be necessary, I can do the rebasing if needed.

**Comment by wesleycho** [comment]
Landed as 22961157ad4636db12336abba54cc893554b99d6, thanks!

**Pull request closed without merging**